src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/Join.cs

```diff
@@ -0,0 +1,237 @@
+            returnType = leftTable.Clone();
+
+            // JoinType argument is present?
+            if (argTypes.Count() > 3 &&
+                (!BuiltInEnums.JoinTypeEnum.FormulaType._type.Accepts(argTypes[3], exact: true, useLegacyDateTimeAccepts: false, usePowerFxV1CompatibilityRules: context.Features.PowerFxV1CompatibilityRules) ||
+                args[3] is not DottedNameNode))
+            {
+                errors.EnsureError(DocumentErrorSeverity.Severe, args[3], TexlStrings.ErrJoinNotPlainJoinTypeEnum);
+                valid = false;
+            }
+
+            ScopeInfo.GetScopeIdent(args, out var scopeIdent);
+
+            if (args.Count() > 4)
+            {                
+                var fError = false;
+
+                // 5th arg on must be renaming/declaration args. It is mandatory for makers to declarer at least 1 RightRecord column.
+                foreach (var node in args.Skip(4))
+                {
+                    if (node is not AsNode asNode)
+                    {
+                        errors.EnsureError(DocumentErrorSeverity.Severe, args[0], TexlStrings.ErrJoinArgIsNotAsNode);
+                        valid = false;
+                        break;
+                    }
+
+                    if (asNode.Left is not DottedNameNode dottedNameNode)
+                    {
+                        errors.EnsureError(DocumentErrorSeverity.Severe, args[0], TexlStrings.ErrJoinArgIsNotAsNode);
+                        valid = false;
+                        break;
+                    }
+                    
+                    if (dottedNameNode.Left.AsFirstName().Ident.Name == scopeIdent[0])
+                    {
+                        if (!leftTable.TryGetType(dottedNameNode.Right.Name, out var leftType))
+                        {
+                            errors.EnsureError(DocumentErrorSeverity.Severe, args[0], TexlStrings.ErrColDNE_Name, dottedNameNode);
+                            valid = false;
+                            break;
+                        }
+
+                        returnType = returnType.Drop(ref fError, DPath.Root, dottedNameNode.Right.Name);
+                        returnType = returnType.Add(ref fError, DPath.Root, asNode.Right.Name, leftType);
+                    }
+                    else if (dottedNameNode.Left.AsFirstName().Ident.Name == scopeIdent[1])
+                    {
+                        // !!TODO shouldn't the binder have caught this?
+                        if (!rightTable.TryGetType(dottedNameNode.Right.Name, out var rightType))
+                        {
+                            errors.EnsureError(DocumentErrorSeverity.Severe, args[0], TexlStrings.ErrColDNE_Name, dottedNameNode);
+                            valid = false;
+                            break;
+                        }
+
+                        returnType = returnType.Add(ref fError, DPath.Root, asNode.Right.Name, rightType);
```

Comment:
>returnType = returnType.Add(ref fError, DPath.Root, asNode.Right.Name, rightType); [](http://example.com/codeflow?start=24&length=82)

We may need to check for duplicates here:

    Join(
      [{a:1,b:2},{a:3,b:4}],
      [{a:1,c:5},{a:3,c:6}],
      LeftRecord.a == RightRecord.a,
      JoinType.Inner,
      RightRecord.c As b
    )

Reply:
`fError` will tell if there are any errors (including duplicates).
This test cases has been added.